lldb/include/lldb/Core/Address.h

```diff
@@ -246,8 +246,8 @@ class Address {
   /// \see Address::DumpStyle
   bool Dump(Stream *s, ExecutionContextScope *exe_scope, DumpStyle style,
             DumpStyle fallback_style = DumpStyleInvalid,
-            uint32_t addr_byte_size = UINT32_MAX,
-            bool all_ranges = false) const;
+            uint32_t addr_byte_size = UINT32_MAX, bool all_ranges = false,
+            const char *pattern = nullptr) const;
```

Comment:
Why is `pattern` a `const char*` and not a StringRef with an empty string as the default argument? Or potentially even better an `std::optional<llvm::Regex>`?

Reply:
Changed to `StringRef`. I could not make it work with `std::optional<llvm::Regex>`. Maybe I was doing something wrong.

Reply:
What errors did you get?

Remember that optionals are "pointer like" when you want to get at the value within them, `*foo`, `foo->method()` etc. that trips me up sometimes. Also the default value would be ` = std::nullopt`.

Reply:
I apologize, I should have noted the errors. But as far as I remember, it was related to use of deleted functions and type conversions. I also tried passing the parameter as a reference thinking the error could be caused by copying. 

> Remember that optionals are "pointer like" when you want to get at the value within them, *foo, foo->method() etc. that trips me up sometimes.

Hmm, I was trying to get `pattern.value()`, perhaps that was part of the problem.

> Also the default value would be  = std::nullopt

At least I got that right.

Reply:
Tbf I had to look up nullopt, I don't see it much. `.value` returns a reference to the contained value (assuming it has one), `->` does the same so either is fine.

Can you post the error message(s)? Don't need to update this change, just describe how you're constructing and passing it, that'll likely be enough.

Reply:
It may be caused by this https://godbolt.org/z/414Kzd18Y because llvm::Regex deletes the copy constructor. This is I think because it contains state about the matching. I think I should be able to move an llvm::Regex though but I haven't been able to get that to work either.

Reply:
Regex is put into an optional [here](https://github.com/llvm/llvm-project/blob/e4710872e98a931c6d5e89bc965b778746ead2c0/mlir/lib/IR/Diagnostics.cpp#L640) and in one other place. See also `clang-tools-extra/clangd/ConfigCompile.cpp`.

Tried to recreate that in a small example but failed. See what you can do with it.

Reply:
> Tbf I had to look up nullopt, I don't see it much.

I search this as well. Actually, I never used `std::optional` before.

> Can you post the error message(s)? Don't need to update the PR, just describe how you're constructing and passing it, that'll likely be enough.

Okay. I'll update the PR with the recent changes and will try to replicate the error and post here.

> It may be caused by this https://godbolt.org/z/414Kzd18Y because llvm::Regex deletes the copy constructor. This is I think because it contains state about the matching. I think I should be able to move an llvm::Regex though but I haven't been able to get that to work either.

This is interesting. I believe this is similar to the error we were getting.

Reply:
Hi @DavidSpickett . Tried this incremental approach:

```
static void test_func1(Stream &strm, llvm::Regex *regex_param = nullptr) {
  strm.Printf("Inside test_func1\n");
}

static void test_func2(Stream &strm, llvm::Regex &regex_param) {
  strm.Printf("Inside test_func2\n");
}

static void test_func3(Stream &strm, const std::optional<llvm::Regex> regex_param = std::nullopt) {
  strm.Printf("Inside test_func3\n");
}

static void test_func4(Stream &strm, const std::optional<llvm::Regex> &regex_param = std::nullopt) {
  strm.Printf("Inside test_func4\n");
}

static uint32_t LookupSymbolInModule(CommandInterpreter &interpreter,
                                     Stream &strm, Module *module,
                                     const char *name, bool name_is_regex,
                                     bool verbose, bool all_ranges) {
  llvm::Regex regex1(name);
  std::optional<llvm::Regex> regex2 = std::make_optional<llvm::Regex>(name);

  test_func1(strm, &regex1);
  test_func2(strm, regex1);
  // test_func3(strm, regex2);
  test_func4(strm, regex2);
  ...
  ...
  }
```

In test1, used the same logic like used previously i.e., function accepting pointer to `llvm::Regex` and default value is null pointer. 
In test2 passed `llvm::regex` by reference but could not give a default value. 
In test 3 used the `std::optional<llvm::Regex>` but was getting the error mentioned above i.e., use of deleted functions. The exact error is:
```
error: use of deleted function ‘std::optional<llvm::Regex>::optional(const std::optional<llvm::Regex>&)’
 1611 |   test_func3(strm, regex2)
```
I believe it's due to the deletion of copy constructor which you mentioned. 
In test 4 passing the reference of `std::optional<llvm::Regex>` to the function which is working fine as expected. Will use this one to pass the pattern now.
I'm a little confused. Right now, we're passing the pattern as `llvm::StringRef`, and finally when we reach `PutCStringColorHighlighted` we convert it to llvm::Regex:
`llvm::Regex reg_pattern(pattern);`
What will be the benefit of converting `const char* name` to `std::optional<llvm::Regex>`  from the very beginning and than passing it till we reach `PutCStringColorHighlighted`? 🤔

Reply:
Hi @DavidSpickett! Sorry for the delay 🙏

`test_func3(strm, std::move(regex2))` works but it gives the following problem. 
When we use `std::move(regex2)` to pass it to `test_func3`, it's essentially moved, and after the move, the `regex2` in `LookupSymbolInModule` becomes garbage.

So in functions where pattern is moved multiple times like this:

```
static void DumpAddress(ExecutionContextScope *exe_scope,
                        const Address &so_addr, bool verbose, bool all_ranges,
                        Stream &strm, std::optional<llvm::Regex> pattern = std::nullopt) {
     ...
     ...
     so_addr.Dump(&strm, exe_scope, Address::DumpStyleResolvedDescription,
               Address::DumpStyleInvalid, UINT32_MAX, false, std::move(pattern));
     ...
     ...
     if (verbose) {
    strm.EOL();
    so_addr.Dump(&strm, exe_scope, Address::DumpStyleDetailedSymbolContext,
                 Address::DumpStyleInvalid, UINT32_MAX, all_ranges, std::move(pattern));
  }
     ...
}
```
The second `std::move(pattern)`, does not move the original pattern value but some garbage value. The only way is to make a copy of the original pattern and then use it with `std::move(copy_pattern)` every time (which doesn't look okay :-/).

This is the example code and output if I couldn't explain it properly above:
CODE:
```
static void test_func3(Stream &strm, const std::optional<llvm::Regex> pattern = std::nullopt) {
  if (pattern.has_value() && pattern.value().match("main")) {
        strm.Printf("Pattern matches 'main' inside test_func3\n");
    }
}

static uint32_t LookupSymbolInModule(CommandInterpreter &interpreter,
                                     Stream &strm, Module *module,
                                     const char *name, bool name_is_regex,
                                     bool verbose, bool all_ranges) {
  // test_func1(strm, std::make_optional<llvm::Regex>(name));
  std::optional<llvm::Regex> regex2 = std::make_optional<llvm::Regex>(name);
  if (regex2.has_value()) {
    strm.Printf("Pattern has a value\n");
    if(regex2.value().match("main")) {
      strm.Printf("Pattern matches main\n");
    }
    else{
      strm.Printf("Pattern doesn't match main\n");
    }
  } else {
    strm.Printf("Pattern doesn't have a value\n");
  }
  test_func3(strm, std::move(regex2));
  if (regex2.has_value()) {
    strm.Printf("Pattern has a value\n");
    if(regex2.value().match("main")) {
      strm.Printf("Pattern matches main\n");
    }
    else{
      strm.Printf("Pattern doesn't match main\n");
    }
  } else {
    strm.Printf("Pattern doesn't have a value\n");
  }  ...
  ...
}
```
OUTPUT:
```
talha@talha-ROG-Strix-G513QE-G513QE:~/Desktop/LLDB/build/bin$ ./lldb a.out 
(lldb) target create "a.out"
Current executable set to '/home/talha/Desktop/LLDB/build/bin/a.out' (x86_64).
(lldb) image lookup -r -s main
Pattern has a value
Pattern matches main
Pattern matches 'main' inside test_func3
Pattern has a value
Pattern doesn't match main
...
...
```
So after `std::move(pattern)`, `pattern != "main"` within the same function 

Similarly passing `std::make_optional<llvm::Regex>(name)` to func3 also causes problem problem:
i.e., `test_func3(strm, std::make_optional<llvm::Regex>(name));`

When we directly pass `std::make_optional<llvm::Regex>(name)`, it creates a temporary `std::optional<llvm::Regex>` object that is passed directly to `test_func3`. This temporary object can be moved without requiring an explicit `std::move` because I guess temporaries are considered rvalues, and `std::optional` has appropriate move constructors to handle them.
But storing the result of `std::make_optional<llvm::Regex>(name)` in a variable and then passing it without using `std::move`, if the variable is passed as an argument to the function, it attempts to perform a copy unless we use `std::move`. So, initially we can pass it directly to the next dumping function but further passing it will require `std::move` and the same problem comes back. 

Also 1 last thing for clarification, as discussed later i.e., 

> I think we may be able to reduce the boilerplate by putting the prefix/suffix/pattern in a struct, and passing that as an optional assuming we can figure out the move issue with regex. But this is for a future change if at all.

The struct will be optional and not the member variables inside it as we can check them in the very first function right? 
So the struct will look like:
```
struct information {
pattern, prefix, sufix
}
```
And information will be passed as optional argument rather than
```
struct information {
optional pattern = EMPTY, optional prefix= EMPTY, optional sufix = EMPTY
}
```

Reply:
>When we use std::move(regex2) to pass it to test_func3, it's essentially moved, and after the move, the regex2 in LookupSymbolInModule becomes garbage.

This is where I get on my Rust powered soapbox and preach about C++ move not being what you'd think it is. :)

Good research anyway. Sound like it's what I suspected is that the regex is carrying around the match "engine"'s state.

On the struct, yes, the struct itself would be optional. If you do pass an instance of it you must provide all 3.

To be honest, once we're packing all this into a struct, that's quite an obvious signal to the programmer that this is no ordinary string. So you could do just try the struct idea without using the `Regex` type.

And ok, it's not perfect because all 3 struct members have the same type still, but if making the pattern a `Regex` is this much trouble it's probably not worth it.

Reply:
So in code that would be roughly:
```
struct information {
pattern, prefix, sufix
}

PutCStringHighlighted(....., std::optional<information> pattern_info);
```

Which tells the reader clearly that if they want to provide any one of the 3 they must provide all 3. Otherwise they must provide none of them.